crates/sled/src/tree.rs

```diff
@@ -288,6 +288,73 @@ impl Tree {
         }
     }
 
+    /// Fetch the value, apply a function to it and return the result.
+    ///
+    /// # Note
+    ///
+    /// This may call the function multiple times if the value has been
+    /// changed from other threads in the meantime.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use sled::{ConfigBuilder, Error};
+    /// let config = ConfigBuilder::new().temporary(true).build();
+    /// let t = sled::Db::start(config).unwrap();
+    ///
+    /// assert_eq!(t.update_and_fetch(b"hello", |_old| Some(vec![1])), Ok(Some(vec![1]));
+    /// ```
+    pub fn update_and_fetch<K, V, F>(&self, key: K, mut f: F) -> Result<Option<IVec>>
+    where K: AsRef<[u8]>,
+          V: Into<IVec>,
+          F: Fn(Option<&[u8]>) -> Option<V>,
+    {
+        let key = key.as_ref();
+        let mut current = self.get(key)?;
+
+        loop {
+            let tmp = current.as_ref().map(AsRef::as_ref);
+            let next = f(tmp).map(Into::into);
+            match self.cas(key, tmp, next.clone())? {
+                Ok(()) => return Ok(next),
+                Err(new_current) => current = new_current,
+            }
+        }
+    }
+
+    /// Fetch the value, apply a function to it and return the previous value.
+    ///
+    /// # Note
+    ///
+    /// This may call the function multiple times if the value has been
+    /// changed from other threads in the meantime.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use sled::{ConfigBuilder, Error};
+    /// let config = ConfigBuilder::new().temporary(true).build();
+    /// let t = sled::Db::start(config).unwrap();
+    ///
+    /// assert_eq!(t.fetch_and_update(b"hello", |_old| Some(vec![1])), Ok(None);
+    /// ```
+    pub fn fetch_and_update<K, V, F>(&self, key: K, mut f: F) -> Result<Option<IVec>>
+    where K: AsRef<[u8]>,
+          V: Into<IVec>,
```

Comment:
Would you mind flipping this to `IVec: From<V>` so we can make it more clear below that we're invoking that specific conversion?